pkg/timeseries/dataset/point.go

```diff
@@ -132,3 +189,69 @@ func (p Points) onOrJustBefore(ts epoch.Epoch, s, e int) int {
 	}
 	return p.onOrJustBefore(ts, s, mid)
 }
+
+// sortAndDedupe sorts and deduplicates p in-place. Because deduplication can
+// shorten p, the version of p used to call sortAndDedupe is no longer valid.
+// Set p to the call's return value, as in:   p = sortAndDedupe(p)
+// In the event of duplicates, the highest index wins.
+func sortAndDedupe(p Points) Points {
+	// sort, keeping order between equal elements
+	slices.SortStableFunc(p, func(a, b Point) int {
+		if a.Epoch < b.Epoch {
+			return -1
+		} else if a.Epoch > b.Epoch {
+			return 1
+		}
+		return 0
+	})
+	var k int
```

Comment:
have you considered: https://pkg.go.dev/slices#CompactFunc ?

Reply:
CompactFunc keeps only the first of consecutive values in the sorted slice, whereas this inline solution keeps only the last element. The idea here is that while the Epochs of two Points being compared could be the same, their underlying values may be different in newer data (which falls on the high/right side of consecutives). And we want to keep the newer data because the cached values could have been preliminary. We could re-arrange some things so the order of the points coming in are newest first, but that could introduce other unforeseen problems, so I'm not sure it's worth it to swap everything out to take advantage of CompactFunc.